core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -386,6 +387,88 @@ public final Boolean readOptionalBoolean() throws IOException {
     @Override
     public abstract int available() throws IOException;
 
+    /**
+     * Uses {@link StreamInput#readVInt()} to first load the size, then it creates an array of the appropriate length using the
+     * {@code arrayCreator}. Then it uses the {@code converter} to fill the array with objects of the specified type.
+     * <p>
+     * This is meant to replace boilerplate surrounding code of the form:
+     * <pre>
+     * YourObject[] objects = new YourObject[in.readVInt()];
+     * for (int i = 0; i &lt; objects.length; i++) {
+     *     objects[i] = new YourObject(in);
+     * }
+     * </pre>
+     * <p>
+     * Do not use this for {@code String}s or primitives (e.g., {@code int}).
+     *
+     * @param arrayCreator Expected to be in the form of {@code YourObject[]::new}.
+     * @param converter Expected to be in the form of {@code YourObject::new} or a static function.
+     * @return Never {@code null}.
+     */
+    public <T> T[] readArray(IntFunction<T[]> arrayCreator, StreamInputFunction<T> converter) throws IOException {
+        int size = readVInt();
+
+        T[] array = arrayCreator.apply(size);
```

Comment:
What about readList? I get that sometimes you have an array and not a list but this seems like overkill.